### PR TITLE
New serialized object, public key, and private key interfaces: 

### PR DESCRIPTION
**Pull request by vinniefalco**
This introduces functions get and set, and a family of specialized structs called `STExchange`. These interfaces allow efficient and seamless interchange between serialized object fields and user defined types, especially variable length objects. 

A new base class template `TypedField` is mixed into existing `SField` declarations to encode information on the field, allowing template metaprograms to both customize interchange based on the type and detect misuse at compile-time.

New types `AnyPublicKey` and `AnySecretKey` are introduced. These are intended to replace the corresponding functionality in the deprecated class `RippleAddress`. Specializations of `STExchange` for these types are provided to allow interchange. New free functions verify and sign allow signature verification and signature generation for serialized objects.

* Add `Buffer` and `Slice` primitives
* Add `TypedField` and modify some `SField`
* Add `STExchange` and specializations for `STBlob` and `STInteger`
* Improve `STBlob` and `STInteger` to support `STExchange`
* Expose raw data in `RippleAddress` and `Serializer`

### CODE REVIEW

File: src/ripple/basics/Slice.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+hash_append (Hasher& h, Slice const& v)
+{
+    h.append(v.data(), v.size());
+}
+
+inline
+bool
+operator== (Slice const& lhs, Slice const& rhs) noexcept
+{
+    return lhs.size() == rhs.size() && std::memcmp(
```
Comment (rec) [inline src/ripple/basics/Slice.h]: This won't compile, as there's an extra trailing `)`!

Comment (vinniefalco) [inline src/ripple/basics/Slice.h]: Should compile now

**Comment by josh-ripple** [comment]
Looks good, :+1: pending a workaround to the secure erase issue.

**Comment by vinniefalco** [comment]
@josh-ripple I changed it to `std::fill`

**Comment by nbougalis** [comment]
I won't comment on something just because the comment asked me to. :+1:. I'll just point out that the `std::fill` isn't really a secure erase - the compiler can (and likely will) optimize it away. We should consider a more comprehensive solution for this issue.

**Comment by rec** [comment]
This causes a unit test failure on OS/X:
`ripple.ripple_app.STTx: #3 failed: Unable to build object from json`

**Comment by vinniefalco** [comment]
@rec Hard to imagine how this change set could cause a unit test failure, because this code introduces no functional changes.

**Pull request closed without merging**